### PR TITLE
Support for FileTypeAssociation registration.

### PR DESCRIPTION
**Pull request by aeloros**
Initial implementation of file type association registrations for Win32.  Also have basic functional tests up and running for both Win32 and Packaged Win32.

The tests now require elevation, in order to install the test apps cert.  Also with this change we switch to TAEF for our test harness.

### CODE REVIEW

File: dev/AppLifecycle/Association.cpp [hunk 1]
```diff
@@ -0,0 +1,423 @@
...
+    std::wstring ComputeAppId()
+    {
+        // PRF = Project Reunion Framework.  A simple prefix to help identify ProgIds we compute here.
+        // AppId = Prefix + Hash(modulePath)
+        auto modulePath = GetModulePath();
+        std::hash<std::wstring> hasher;
+        auto hash = hasher(modulePath);
+;
+        wchar_t hashString[17]{}; // 16 + 1 characters for 64bit value represented as a string with a null terminator.
+        THROW_IF_FAILED(StringCchPrintf(hashString, sizeof(hashString), L"%I64x", hash));
```
Comment (EHO-makai) [inline dev/AppLifecycle/Association.cpp:66]: StringCchPrintf is "char count" based, sizeof is byte based.  Is this combo correct?

File: dev/AppLifecycle/Association.cpp [hunk 2]
```diff
@@ -0,0 +1,423 @@
...
+            KEY_WRITE, nullptr, key.put(), nullptr));
+
+        THROW_IF_WIN32_ERROR(::RegSetValueEx(key.get(), nullptr, 0, REG_SZ,
+            reinterpret_cast<BYTE const*>(command.c_str()),
+            static_cast<uint32_t>((command.size() + 1) * sizeof(wchar_t))));
+
+        if (delegateExecute)
+        {
+            std::wstring delegateClsid{ LR"({????????-????-????-????-????????????})" };
+            ::StringFromGUID2(*delegateExecute, delegateClsid.data(), 39);
```
Comment (EHO-makai) [inline dev/AppLifecycle/Association.cpp:273]: possible to do some sizeof(guid) or equivalent?

File: dev/AppLifecycle/Association.cpp [hunk 3]
```diff
@@ -0,0 +1,423 @@
...
+    }
+
+    void UnregisterVerb(const std::wstring& progId, const std::wstring& verb)
+    {
+        auto key_path = CreateAssocKeyPath(progId) + L"\\" + c_shellKeyName + L"\\" + verb +
+            L"\\" + c_commandKeyName;
+        ::RegDeleteTree(GetRegistrationRoot(), key_path.c_str());
+    }
+
+    void RegisterProtocol(const std::wstring& scheme)
```
Comment (EHO-makai) [inline dev/AppLifecycle/Association.cpp:288]: in many of the functions in this section, I find myself wanting a short comment that shows an example of the string that is being built.

File: dev/AppLifecycle/FileActivatedEventArgs.h [hunk 4]
```diff
@@ -0,0 +1,53 @@
...
+{
+    using namespace winrt::Windows::Foundation::Collections;
+    using namespace winrt::Windows::ApplicationModel::Activation;
+    using namespace winrt::Windows::Storage;
+
+    class FileActivatedEventArgs : public winrt::implements<FileActivatedEventArgs,
+        ActivatedEventArgsBase, IFileActivatedEventArgs>
+    {
+    public:
+        FileActivatedEventArgs([[maybe_unused]] const std::wstring contractData)
```
Comment (EHO-makai) [inline dev/AppLifecycle/FileActivatedEventArgs.h:17]: seems like this isn't necessary?

File: dev/AppLifecycle/LaunchActivatedEventArgs.h [hunk 5]
```diff
@@ -12,8 +12,7 @@ namespace winrt::Microsoft::ProjectReunion::implementation
         ActivatedEventArgsBase, ILaunchActivatedEventArgs>
     {
     public:
-        LaunchActivatedEventArgs(const std::wstring& args) :
-            m_args(args)
+        LaunchActivatedEventArgs(const std::wstring& args) : m_args(args)
```
Comment (EHO-makai) [inline dev/AppLifecycle/LaunchActivatedEventArgs.h:15]: thank you

File: test/TestApps/AppLifecycleTestApp/main.cpp [hunk 6]
```diff
@@ -0,0 +1,131 @@
...
+        }
+    }
+    else if (kind == ActivationKind::File)
+    {
+        // Validate access to the files on the arguments.
+        auto fileArgs = args.as<IFileActivatedEventArgs>();
+        for (auto const& item : fileArgs.Files())
+        {
+            auto file = item.as<StorageFile>();
+            auto stream = file.OpenAsync(FileAccessMode::Read).get();
```
Comment (EHO-makai) [inline test/TestApps/AppLifecycleTestApp/main.cpp:113]: how does this work?  do you have to co_await here?

Comment (EHO-makai) [reply test/TestApps/AppLifecycleTestApp/main.cpp:113]: Sorry, I missed the .get() at the end of the line.

File: dev/AppLifecycle/Association.cpp (hunk 1 above)
Comment (aeloros) [reply dev/AppLifecycle/Association.cpp:66]: Doh, definitely not correct.  Good catch!

File: dev/AppLifecycle/FileActivatedEventArgs.h (hunk 4 above)
Comment (aeloros) [reply dev/AppLifecycle/FileActivatedEventArgs.h:17]: I assume you mean the [[maybe_unused]]?  Yeah, that is old.  I'll remove.  Thanks!

File: test/AppLifecycle/FunctionalTests.cpp [hunk 7]
```diff
@@ -0,0 +1,217 @@
...
+            VERIFY_IS_NOT_NULL(launchArgs);
+        }
+
+        TEST_METHOD(GetActivatedEventArgsForProtocol_Win32)
+        {
+            // Create a named event for communicating with test app.
+            auto event = CreateTestEvent(c_testProtocolPhaseEventName);
+
+            // Launch the test app to register for protocol launches.
+            if (!Execute(L"AppLifecycleTestApp.exe", L"/RegisterProtocol", g_deploymentDir))
```
Comment (dabhattimsft) [inline test/AppLifecycle/FunctionalTests.cpp:131]: nit: May be just do VERIFY_NOT_NULL(Execute(...));? There is already a GetLastError in Execute() to get error if ShellExecute fails.

**Review by dabhattimsft — APPROVED** [review]

**Review by EHO-makai — APPROVED** [review]

File: test/AppLifecycle/FunctionalTests.cpp (hunk 7 above)
Comment (aeloros) [reply test/AppLifecycle/FunctionalTests.cpp:131]: Good catch!  I wrote these before updating the wrapper for the cert util calls, and forgot to update all the callers.

**Pull request merged**